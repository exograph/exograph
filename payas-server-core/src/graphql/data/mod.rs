pub mod data_root_element;

mod data_operation;
mod interception;
mod mutation_resolver;
mod operation_resolver;
mod query_resolver;
