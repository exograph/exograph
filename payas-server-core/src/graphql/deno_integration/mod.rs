pub mod clay_execution;
pub mod claytip_ops;

pub use clay_execution::{
    clay_config, ClayCallbackProcessor, FnClaytipExecuteQuery, FnClaytipInterceptorProceed,
};
pub use claytip_ops::InterceptedOperationInfo;
use payas_deno::DenoExecutorPool;
pub type ClayDenoExecutorPool = DenoExecutorPool<
    Option<InterceptedOperationInfo>,
    clay_execution::RequestFromDenoMessage,
    clay_execution::ClaytipMethodResponse,
>;
