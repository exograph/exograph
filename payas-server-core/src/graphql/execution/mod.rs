pub mod field_resolver;
mod root_operation_resolver;
pub mod system_context;
