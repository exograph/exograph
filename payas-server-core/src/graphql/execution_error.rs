use std::error::Error;

use payas_resolver_database::DatabaseExecutionError;
use payas_resolver_wasm::WasmExecutionError;
use thiserror::Error;

use crate::graphql::validation::validation_error::ValidationError;

use payas_resolver_deno::DenoExecutionError;

#[derive(Error, Debug)]
pub enum ExecutionError {
    #[error("{0}")]
    Generic(String),

    #[error("{0}")]
    Database(#[from] DatabaseExecutionError),

    #[error("{0}")]
    Deno(#[from] DenoExecutionError),

    #[error("{0}")]
    Wasm(#[from] WasmExecutionError),

    #[error("{0}")]
    Serde(#[from] serde_json::Error),

    #[error("{0}")]
    Validation(#[from] ValidationError),

    #[error("Invalid field {0} for {1}")]
    InvalidField(String, &'static str), // (field name, container type)
}

impl ExecutionError {
    // Message that should be emitted when the error is returned to the user.
    // This should hide any internal details of the error.
    // TODO: Log the details of the error.
    pub fn user_error_message(&self) -> String {
        match self {
            ExecutionError::Database(error) => error.user_error_message(),
            ExecutionError::Deno(error) => match error {
                DenoExecutionError::Delegate(underlying) => {
                    match underlying.downcast_ref::<ExecutionError>() {
                        Some(error) => error.user_error_message(),
                        None => error.user_error_message(),
                    }
                }
                _ => error.user_error_message(),
            },
            _ => match self.source() {
                Some(source) => source.to_string(),
                None => self.to_string(),
            },
        }
    }
}
