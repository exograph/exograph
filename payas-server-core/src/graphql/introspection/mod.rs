pub mod definition;
pub mod resolver;
