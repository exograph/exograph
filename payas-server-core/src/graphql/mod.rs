pub mod data;

pub mod execution;
pub mod execution_error;
pub mod introspection;
pub mod request_context;
mod validation;
