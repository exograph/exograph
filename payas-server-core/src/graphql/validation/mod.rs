use payas_model::model::system::ModelSystem;

use self::definition::GqlTypeDefinition;

/// Validate the query payload.
///
/// Take a user submitted query along with the operation name and variables (from the request payload)
/// and transform the query into a validated form (in the process, validate the query).
pub mod operation;

pub mod document_validator;

mod arguments_validator;
mod definition;
mod gql_field_definition;
mod gql_parameter_definition;
mod gql_type_definition;
mod operation_validator;
mod selection_set_validator;
pub mod validation_error;

fn find_arg_type<'a>(model: &'a ModelSystem, name: &str) -> Option<&'a dyn GqlTypeDefinition> {
    if let Some(typ) = model
        .predicate_types
        .iter()
        .find(|t| t.1.name.as_str() == name)
        .map(|t| t.1 as &dyn GqlTypeDefinition)
    {
        return Some(typ);
    }

    if let Some(typ) = model
        .order_by_types
        .iter()
        .find(|t| t.1.name.as_str() == name)
        .map(|t| t.1 as &dyn GqlTypeDefinition)
    {
        return Some(typ);
    }

    if let Some(typ) = model
        .mutation_types
        .iter()
        .find(|t| t.1.name.as_str() == name)
        .map(|t| t.1 as &dyn GqlTypeDefinition)
    {
        return Some(typ);
    }

    if let Some(typ) = model
        .argument_types
        .iter()
        .find(|t| t.1.name.as_str() == name)
        .map(|t| t.1 as &dyn GqlTypeDefinition)
    {
        return Some(typ);
    }

    None
}
