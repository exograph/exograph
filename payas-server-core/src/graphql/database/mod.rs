pub mod cast;
mod create_data_param_mapper;
mod limit_offset_mapper;

mod order_by_mapper;
pub mod predicate_mapper;

pub mod database_query;
pub mod sql_mapper;
mod update_data_param_mapper;

use std::collections::HashMap;

use postgres_types::FromSqlOwned;
use predicate_mapper::PredicateParameterMapper;

use crate::graphql::validation::field::ValidatedField;
use async_graphql_value::ConstValue;
use tokio_postgres::Row;

use sql_mapper::{SQLInsertMapper, SQLUpdateMapper};

use payas_sql::{
    AbstractDelete, AbstractInsert, AbstractPredicate, AbstractSelect, AbstractUpdate, ColumnPath,
    ColumnPathLink, PhysicalColumn, PhysicalTable,
};

use crate::graphql::{
    execution::system_context::SystemContext,
    execution_error::{ExecutionError, WithContext},
    request_context::RequestContext,
};

use payas_model::model::{
    column_id::ColumnId,
    operation::{CreateDataParameter, Mutation, OperationReturnType, Query, UpdateDataParameter},
    predicate::{ColumnIdPath, ColumnIdPathLink, PredicateParameter},
    system::ModelSystem,
    GqlCompositeType, GqlTypeKind,
};

use self::sql_mapper::SQLOperationKind;

use super::{data::access_solver, execution_error::DatabaseExecutionError};

pub type Arguments = HashMap<String, ConstValue>;

pub async fn compute_sql_access_predicate<'a>(
    return_type: &OperationReturnType,
    kind: &SQLOperationKind,
    system_context: &'a SystemContext,
    request_context: &'a RequestContext<'a>,
) -> AbstractPredicate<'a> {
    let return_type = return_type.typ(&system_context.system);

    match &return_type.kind {
        GqlTypeKind::Primitive => AbstractPredicate::True,
        GqlTypeKind::Composite(GqlCompositeType { access, .. }) => {
            let access_expr = match kind {
                SQLOperationKind::Create => &access.creation,
                SQLOperationKind::Retrieve => &access.read,
                SQLOperationKind::Update => &access.update,
                SQLOperationKind::Delete => &access.delete,
            };
            access_solver::solve_access(access_expr, request_context, &system_context.system).await
        }
    }
}

fn find_arg<'a>(arguments: &'a Arguments, arg_name: &str) -> Option<&'a ConstValue> {
    arguments.iter().find_map(|argument| {
        let (argument_name, argument_value) = argument;
        if arg_name == argument_name {
            Some(argument_value)
        } else {
            None
        }
    })
}

fn compute_predicate<'a>(
    predicate_param: Option<&'a PredicateParameter>,
    arguments: &'a Arguments,
    additional_predicate: AbstractPredicate<'a>,
    system_context: &'a SystemContext,
) -> Result<AbstractPredicate<'a>, ExecutionError> {
    let mapped = predicate_param
        .as_ref()
        .and_then(|predicate_parameter| {
            let argument_value = find_arg(arguments, &predicate_parameter.name);
            argument_value.map(|argument_value| {
                predicate_parameter.map_to_predicate(argument_value, None, system_context)
            })
        })
        .transpose()
        .with_context("While mapping predicate parameters to SQL".into())?;

    let res = match mapped {
        Some(predicate) => {
            AbstractPredicate::And(Box::new(predicate), Box::new(additional_predicate))
        }
        None => additional_predicate,
    };

    Ok(res)
}

pub fn to_column_id_path(
    parent_column_id_path: &Option<ColumnIdPath>,
    next_column_id_path_link: &Option<ColumnIdPathLink>,
) -> Option<ColumnIdPath> {
    match (parent_column_id_path, next_column_id_path_link) {
        (Some(parent_column_id_path), Some(next_column_id_path_link)) => {
            let mut path: Vec<_> = parent_column_id_path.path.clone();
            path.push(next_column_id_path_link.clone());
            Some(ColumnIdPath { path })
        }
        (Some(parent_column_id_path), None) => Some(parent_column_id_path.clone()),
        (None, Some(next_column_id_path_link)) => Some(ColumnIdPath {
            path: vec![next_column_id_path_link.clone()],
        }),
        (None, None) => None,
    }
}

fn to_column_table(column_id: ColumnId, system: &ModelSystem) -> (&PhysicalColumn, &PhysicalTable) {
    let column = column_id.get_column(system);
    let table = &system
        .tables
        .iter()
        .find(|(_, table)| table.name == column.table_name)
        .map(|(_, table)| table)
        .unwrap_or_else(|| panic!("Table {} not found", column.table_name));

    (column, table)
}

fn to_column_path_link<'a>(link: &ColumnIdPathLink, system: &'a ModelSystem) -> ColumnPathLink<'a> {
    ColumnPathLink {
        self_column: to_column_table(link.self_column_id, system),
        linked_column: link
            .linked_column_id
            .map(|linked_column_id| to_column_table(linked_column_id, system)),
    }
}

pub fn to_column_path<'a>(
    parent_column_id_path: &Option<ColumnIdPath>,
    next_column_id_path_link: &Option<ColumnIdPathLink>,
    system: &'a ModelSystem,
) -> ColumnPath<'a> {
    let mut path: Vec<_> = match parent_column_id_path {
        Some(parent_column_id_path) => parent_column_id_path
            .path
            .iter()
            .map(|link| to_column_path_link(link, system))
            .collect(),
        None => vec![],
    };

    if let Some(next_column_id_path_link) = next_column_id_path_link {
        path.push(to_column_path_link(next_column_id_path_link, system));
    }

    ColumnPath::Physical(path)
}

pub fn get_argument_field<'a>(
    argument_value: &'a ConstValue,
    field_name: &str,
) -> Option<&'a ConstValue> {
    match argument_value {
        ConstValue::Object(value) => value.get(field_name),
        _ => None,
    }
}

pub fn extractor<T: FromSqlOwned>(row: Row) -> Result<T, DatabaseExecutionError> {
    match row.try_get(0) {
        Ok(col) => Ok(col),
        Err(err) => Err(DatabaseExecutionError::EmptyRow(err)),
    }
}

pub async fn create_operation<'a>(
    mutation: &'a Mutation,
    data_param: &'a CreateDataParameter,
    field: &'a ValidatedField,
    select: AbstractSelect<'a>,
    system_context: &'a SystemContext,
    request_context: &'a RequestContext<'a>,
) -> Result<AbstractInsert<'a>, ExecutionError> {
    // TODO: https://github.com/payalabs/payas/issues/343
    let access_predicate = compute_sql_access_predicate(
        &mutation.return_type,
        &SQLOperationKind::Create,
        system_context,
        request_context,
    )
    .await;

    // TODO: Allow access_predicate to have a residue that we can evaluate against data_param
    // See issue #69
    if access_predicate == AbstractPredicate::False {
        // Hard failure, no need to proceed to restrict the predicate in SQL
        return Err(ExecutionError::Authorization);
    }

    let argument_value = find_arg(&field.arguments, &data_param.name).unwrap();

    data_param.insert_operation(mutation, select, argument_value, system_context)
}

pub async fn delete_operation<'a>(
    mutation: &'a Mutation,
    predicate_param: &'a PredicateParameter,
    field: &'a ValidatedField,
    select: AbstractSelect<'a>,
    system_context: &'a SystemContext,
    request_context: &'a RequestContext<'a>,
) -> Result<AbstractDelete<'a>, ExecutionError> {
    let (table, _, _) = return_type_info(mutation, system_context);

    // TODO: https://github.com/payalabs/payas/issues/343
    let access_predicate = compute_sql_access_predicate(
        &mutation.return_type,
        &SQLOperationKind::Delete,
        system_context,
        request_context,
    )
    .await;

    if access_predicate == AbstractPredicate::False {
        // Hard failure, no need to proceed to restrict the predicate in SQL
        return Err(ExecutionError::Authorization);
    }

    let predicate = compute_predicate(
        Some(predicate_param),
        &field.arguments,
        AbstractPredicate::True,
        system_context,
    )
    .with_context(format!(
        "During predicate computation for parameter {}",
        predicate_param.name
    ))?;

    Ok(AbstractDelete {
        table,
        predicate: Some(predicate),
        selection: select,
    })
}

pub async fn update_operation<'a>(
    mutation: &'a Mutation,
    data_param: &'a UpdateDataParameter,
    predicate_param: &'a PredicateParameter,
    field: &'a ValidatedField,
    select: AbstractSelect<'a>,
    system_context: &'a SystemContext,
    request_context: &'a RequestContext<'a>,
) -> Result<AbstractUpdate<'a>, ExecutionError> {
    // Access control as well as predicate computation isn't working fully yet. Specifically,
    // nested predicates aren't working.
    // TODO: https://github.com/payalabs/payas/issues/343
    let access_predicate = compute_sql_access_predicate(
        &mutation.return_type,
        &SQLOperationKind::Update,
        system_context,
        request_context,
    )
    .await;

    if access_predicate == AbstractPredicate::False {
        // Hard failure, no need to proceed to restrict the predicate in SQL
        return Err(ExecutionError::Authorization);
    }

    // TODO: https://github.com/payalabs/payas/issues/343
    let predicate = compute_predicate(
        Some(predicate_param),
        &field.arguments,
        AbstractPredicate::True,
        system_context,
    )
    .with_context(format!(
        "During predicate computation for parameter {}",
        predicate_param.name
    ))?;

    let argument_value = find_arg(&field.arguments, &data_param.name);
    argument_value
        .map(|argument_value| {
            data_param.update_operation(mutation, predicate, select, argument_value, system_context)
        })
        .unwrap()
}

///
/// # Returns
/// - A (table associated with the return type, pk query, collection query) tuple.
pub fn return_type_info<'a>(
    mutation: &'a Mutation,
    system_context: &'a SystemContext,
) -> (&'a PhysicalTable, &'a Query, &'a Query) {
    let system = &system_context.system;
    let typ = mutation.return_type.typ(system);

    match &typ.kind {
        GqlTypeKind::Primitive => panic!(""),
        GqlTypeKind::Composite(kind) => (
            &system.tables[kind.get_table_id()],
            &system.queries[kind.get_pk_query()],
            &system.queries[kind.get_collection_query()],
        ),
    }
}
