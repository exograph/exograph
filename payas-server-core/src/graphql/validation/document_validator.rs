use async_graphql_parser::types::{DocumentOperations, ExecutableDocument};
use async_graphql_value::Name;
use payas_model::model::system::ModelSystem;
use serde_json::{Map, Value};
use tracing::instrument;

use crate::graphql::{
    introspection::definition::schema::Schema, validation::validation_error::ValidationError,
};

use super::{operation::ValidatedOperation, operation_validator::OperationValidator};

/// Context for validating a document.
pub struct DocumentValidator<'a> {
    system: &'a ModelSystem,
    schema: &'a Schema,
    operation_name: Option<String>,
    variables: Option<Map<String, Value>>,
}

impl<'a> DocumentValidator<'a> {
    pub fn new(
        system: &'a ModelSystem,
        schema: &'a Schema,
        operation_name: Option<String>,
        variables: Option<Map<String, Value>>,
    ) -> Self {
        Self {
            system,
            schema,
            operation_name,
            variables,
        }
    }

    /// Validate the query payload.
    ///
    /// Validations performed:
    /// - Validate that either there is only one operation or the operation name specified matches one of the operations in the document
    /// - Validate that there is at least one operation
    /// - Other validations are delegated to the operation validator
    #[instrument(
        name = "DocumentValidator::validate"
        skip(self, document)
        )]
    pub fn validate(
        self,
        document: ExecutableDocument,
    ) -> Result<ValidatedOperation, ValidationError> {
        let (operation_name, raw_operation) = match document.operations {
            DocumentOperations::Single(operation) => Ok((self.operation_name, operation)),
            DocumentOperations::Multiple(mut operations) => {
                if operations.is_empty() {
                    Err(ValidationError::NoOperationFound)
                } else {
                    match self.operation_name {
                        None if operations.len() == 1 => {
                            // Per https://graphql.org/learn/queries/#operation-name, `operationName` is required
                            // only for multiple operations, but async-graphql parses a named operation (`query Foo { ... }`)
                            // to `DocumentOperations::Multiple` even if there is only one operation. So we add an additional
                            // check here to make sure that the operation name is enforced only for truly multiple operations.

                            // This unwrap is okay because we already check that there is exactly one operation.
                            let (operation_name, operation) =
                                operations.into_iter().next().unwrap();
                            Ok((Some(operation_name.to_string()), operation))
                        }
                        None => Err(ValidationError::MultipleOperationsNoOperationName),
                        Some(operation_name) => {
                            let operation = operations.remove(&Name::new(&operation_name));

                            match operation {
                                None => {
                                    Err(ValidationError::MultipleOperationsUnmatchedOperationName(
                                        operation_name,
                                    ))
                                }
                                Some(operation) => Ok((Some(operation_name), operation)),
                            }
                        }
                    }
                }
            }
        }?;

        let operation_validator = OperationValidator::new(
            self.system,
            self.schema,
            operation_name,
            self.variables,
            document.fragments,
        );

        operation_validator.validate(raw_operation)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use async_graphql_parser::parse_query;

    // Macro to avoid insta-related issues with the test names when asserted through a helper function
    // It accepts arguments in the same way as that would be passed to the graphql endpoint
    macro_rules! assert_validation {
        ($query:expr, $operation_name:expr, $variables:expr) => {
            let system = create_test_system();
            let schema = Schema::new(&system);
            let variables = if $variables.is_empty() {
                None
            } else {
                Some(create_variables($variables))
            };

            let operation_name = if $operation_name.is_empty() {
                None
            } else {
                Some($operation_name.to_string())
            };

            let validator = DocumentValidator {
                system: &system,
                schema: &schema,
                operation_name,
                variables,
            };

            insta::assert_debug_snapshot!(validator.validate(create_query_document($query)));
        };
    }

    #[test]
    fn no_arguments_valid() {
        assert_validation!(
            r#"
                query {
                    concerts {
                        id
                        title
                        venue {
                            id
                            name
                        }
                    }
                }
            "#,
            "",
            ""
        );
    }

    #[test]
    fn primitive_argument_valid() {
        assert_validation!(
            r#"
                query {
                    concert(id: 1) {
                        id
                        title
                        venue {
                            id
                            name
                        }
                    }
                }
            "#,
            "",
            ""
        );
    }

    #[test]
    fn composite_argument_valid() {
        assert_validation!(
            r#"
                query {
                    concerts(where: {id: {eq: 1}}) {
                        id
                        title
                        venue {
                            id
                            name
                        }
                    }
                }
            "#,
            "",
            ""
        );
    }

    #[test]
    fn composite_list_argument_valid() {
        assert_validation!(
            r#"
                query {
                    concerts(where: {and: [{id: {gt: 1}}, {id: {lt: 3}}]}) {
                        id
                        title
                        venue {
                            id
                            name
                        }
                    }
                }
            "#,
            "",
            ""
        );
    }

    #[test]
    fn composite_order_by_valid_implicit_array() {
        assert_validation!(
            r#"
                query {
                    concerts(orderBy: {title: DESC}) {
                        id
                    }
                }
            "#,
            "",
            ""
        );
    }

    #[test]
    fn composite_order_by_valid_explicit_array() {
        assert_validation!(
            r#"
                query {
                    concerts(orderBy: [{id: ASC}, {title: DESC}]) {
                        id
                    }
                }
            "#,
            "",
            ""
        );
    }

    #[test]
    fn composite_order_by_invalid_overly_nested_array() {
        assert_validation!(
            r#"
                query {
                    concerts(orderBy: [[{id: ASC}, {title: DESC}]]) {
                        id
                    }
                }
            "#,
            "",
            ""
        );
    }

    #[test]
    fn with_operation_name_valid() {
        assert_validation!(
            r#"
                query ConcertById {
                    concert(id: 1) {
                        id
                        title
                        venue {
                            id
                            name
                        }
                    }
                }
            "#,
            "",
            ""
        );
    }

    #[test]
    fn stray_argument_invalid() {
        assert_validation!(
            r#"
                query {
                    concert(id: 1, foo: "bar") { # foo is not a valid argument
                        id
                        title
                        venue {
                            id
                            name
                        }
                    }
                }
            "#,
            "",
            ""
        );
    }

    #[test]
    fn unspecified_required_argument_invalid() {
        assert_validation!(
            r#"
                query {
                    concert { # id argument is required here
                        id
                        title
                        venue {
                            id
                            name
                        }
                    }
                }
            "#,
            "",
            ""
        );
    }

    #[test]
    fn variable_resolution_valid() {
        assert_validation!(
            r#"
                query($concert_id: Int!, $venue_id: Int!) {
                    concert(id: $concert_id) { # id argument is required here
                        id
                        title
                    }
                    venue(id: $venue_id) {
                        id
                        name
                    }
                }
            "#,
            "",
            r#"
            {
                "concert_id": 1,
                "venue_id": 2
            }"#
        );
    }

    #[test]
    fn variable_resolution_invalid() {
        assert_validation!(
            r#"
            query($concert_id: Int!, $venue_id: Int!) { # venue_id is not a specified in variables
                concert(id: $concert_id) { # id argument is required here
                    id
                    title
                }
                venue(id: $venue_id) {
                    id
                    name
                }
            }
        "#,
            "",
            r#"{ "concert_id": 2 }"#
        );
    }

    #[test]
    fn invalid_subfield() {
        assert_validation!(
            r#"
                query {
                    concert(id: 1) {
                        id
                        title
                        foobar
                    }
                }
            "#,
            "",
            ""
        );
    }

    #[test]
    fn aliases_valid() {
        assert_validation!(
            r#"
                query {
                    firstConcert: concert(id: 1) {
                        id
                        headLine: title
                    }
                }
            "#,
            "",
            ""
        );
    }

    #[test]
    fn multi_operations_valid() {
        let query = r#"
            query concert1 {
                concert(id: 1) {
                    id
                    headLine: title
                }
            }

            query concert2 {
                concert(id: 2) {
                    id
                    headLine: title
                }
            }
        "#;
        assert_validation!(query, "concert1", "");
        assert_validation!(query, "concert2", "");
    }

    #[test]
    fn multi_operations_no_operation_name_invalid() {
        assert_validation!(
            r#"
                query concert1 {
                    concert(id: 1) {
                        id
                        headLine: title
                    }
                }

                query concert2 {
                    concert(id: 2) {
                        id
                        headLine: title
                    }
                }
            "#,
            "",
            ""
        );
    }

    #[test]
    fn multi_operations_mismatched_operation_name_invalid() {
        assert_validation!(
            r#"
                query concert1 {
                    concert(id: 1) {
                        id
                        headLine: title
                    }
                }

                query concert2 {
                    concert(id: 2) {
                        id
                        headLine: title
                    }
                }
            "#,
            "foo",
            ""
        );
    }

    #[test]
    fn basic_mutation() {
        assert_validation!(
            r#"
                mutation {
                    createConcert(data: {title: "Concert 1", venue: {id: 1}}) {
                        id
                        title
                    }
                }
            "#,
            "",
            ""
        );
    }

    #[test]
    fn create_mutation_list_input() {
        assert_validation!(
            r#"
                mutation {
                    createConcerts(data: [{title: "Concert 1", venue: {id: 1}}, {title: "Concert 2", venue: {id: 2}}]) {
                        id
                        title
                    }
                }
            "#,
            "",
            ""
        );
    }

    #[test]
    fn create_mutation_invalid_overly_nested_list_input() {
        assert_validation!(
            r#"
                mutation {
                    createConcerts(data: [[{title: "Concert 1", venue: {id: 1}}]]) {
                        id
                        title
                    }
                }
            "#,
            "",
            ""
        );
    }

    #[test]
    fn service_mutation() {
        assert_validation!(
            r#"
                mutation {
                    processInfo(info: { level: 1, message: "allowed" })
                }
            "#,
            "",
            ""
        );
    }

    fn create_variables(variables: &str) -> Map<String, Value> {
        serde_json::from_str(variables).unwrap()
    }

    fn create_test_system() -> ModelSystem {
        let test_clay = r#"
            model Concert {
                id: Int = autoincrement() @pk
                title: String
                venue: Venue
            }

            model Venue {
                id: Int = autoincrement() @pk
                name: String
                concerts: Set<Concert>
            }

            @external("test.js")
            service TestService {
                input type Info {
                    level: Int
                    message: String
                }
            
                mutation processInfo(info: Info): Boolean
            }
        "#;
        payas_parser::build_system_from_str(test_clay, "test.clay".to_string()).unwrap()
    }

    fn create_query_document(query_str: &str) -> ExecutableDocument {
        parse_query(query_str).unwrap()
    }
}
