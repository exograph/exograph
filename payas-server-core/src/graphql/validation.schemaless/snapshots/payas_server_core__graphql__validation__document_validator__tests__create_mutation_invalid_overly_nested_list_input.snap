---
source: payas-server-core/src/graphql/validation/document_validator.rs
expression: "validator.validate(create_query_document(r#\"\n                mutation {\n                    createConcerts(data: [[{title: \"Concert 1\", venue: {id: 1}}]]) {\n                        id\n                        title\n                    }\n                }\n            \"#))"
---
Err(
    InvalidArgumentType {
        argument_name: "data",
        expected_type: "ConcertCreationInput",
        actual_type: "[ConcertCreationInput]",
        pos: Pos(3:42),
    },
)
